pkg/util/grpc/encoding/snappy/snappy.go

```diff
@@ -0,0 +1,81 @@
+package snappy
+
+import (
+	"io"
+	"sync"
+
+	"github.com/golang/snappy"
+	"google.golang.org/grpc/encoding"
+)
+
+// Name is the name registered for the snappy compressor.
+const Name = "snappy"
+
+func init() {
+	encoding.RegisterCompressor(newCompressor())
+}
+
+type compressor struct {
+	writersPool sync.Pool
+	readersPool sync.Pool
+}
+
+func newCompressor() *compressor {
+	c := &compressor{}
+	c.readersPool = sync.Pool{
+		New: func() interface{} {
+			return &reader{
+				compressor: c,
+				Reader:     snappy.NewReader(nil),
+			}
+		},
+	}
+	c.writersPool = sync.Pool{
+		New: func() interface{} {
+			return &writeCloser{
+				compressor: c,
+				Writer:     snappy.NewBufferedWriter(nil),
+			}
+		},
+	}
+	return c
+}
+
+func (c *compressor) Name() string {
+	return Name
+}
+
+func (c *compressor) Compress(w io.Writer) (io.WriteCloser, error) {
+	wr := c.writersPool.Get().(*writeCloser)
+	wr.Reset(w)
+	return wr, nil
+}
+
+func (c *compressor) Decompress(r io.Reader) (io.Reader, error) {
+	dr := c.readersPool.Get().(*reader)
+	dr.Reset(r)
+	return dr, nil
+}
+
+type writeCloser struct {
+	*compressor
+	*snappy.Writer
```

Comment:
I'd prefer to avoid nameless embeds. They expose more methods than intended (writeCloser now has methods like `Name` and `Compress`), and may inadvertently implement methods that were not meant to be implemented.